### ISSUE TITLE
POST "images/create" can silently fail under certain circumstances

### ISSUE THREAD

**Issue by witlessbird** [docker/docker#11176]
If using a misconfigured external registry, it is possible for POST "/images/create" to fail, but returned http status will be 2xx.

This can happen when TagStore.pullRepository unexpectedly fails on Session.GetRepositoryData call here: https://github.com/docker/docker/blob/master/graph/pull.go#L112.

line at https://github.com/docker/docker/blob/master/api/server/server.go#L594 evaluates to false, and postImagesCreate returns a nil, which in turn, results in httpHandler (https://github.com/docker/docker/blob/master/api/server/server.go#L1244) not returning an error status code.

a failing test to demonstrate the issue: 
```
func TestImagesCreateFailure(t *testing.T) {
        eng := engine.New()
        var called bool
        eng.Register("pull", func(job *engine.Job) engine.Status {
                called = true
                job.Stdout.Write([]byte(""))
                return engine.StatusErr
        })

        data := url.Values{}
        data.Set("fromImage", "test")
        r := httptest.NewRecorder()
        req, err := http.NewRequest("POST", "/images/create", bytes.NewBufferString(data.Encode()))
        if err != nil {
                t.Fatal(err)
        }
        req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
        req.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
        ServeRequest(eng, api.APIVERSION, r, req)

        if !called {
                t.Fatalf("handler was not called")
        }
        if r.Code == http.StatusOK {
                t.Fatal("http status shouldn't be 2xx")
        }
}
```

**Comment by cpuguy83** [comment]
@witlessbird You seem to have a handle on this, care to make a PR?

**Comment by witlessbird** [comment]
@cpuguy83: Sure, I'll give it a shot. Any reason for the logic inside https://github.com/docker/docker/blob/master/api/server/server.go#L593? As in why not return an error if stdout is dirty?

**Comment by cpuguy83** [comment]
Not sure, looks like it goes back a long way.

**Comment by witlessbird** [comment]
k, cheers. I'll take a look around.

**Comment by witlessbird** [comment]
In order to fix this issue, a design decision regarding handling of job stdout has to be made/revised. 

Currently postImagesCreate (https://github.com/docker/docker/blob/master/api/server/server.go#L589) writes and flushes job's output to http.ResponseWriter associated with the call. According to http module docs, the first call to http.ResponseWriter.Write sets the status to http.StatusOK, unless the status has been set already.

 In other words, in postImagesCreate and a few other calls, it's impossible to return any status code other than 200, once job.Stdout has been written to.

Fixes I can think of would have an impact on api clients and/or affect memory usage, so I feel this should probably be handled by you, guys. Having said that, lemme know if I can lend you a hand.

**Comment by witlessbird** [comment]
I created a PR with one of the possible ways to fix the issue: https://github.com/docker/docker/pull/11375
